File: frontend/js/views/irmaa.js
// IRMAA — Medicare Part B/D income surcharge (2026). Maps a two-year-prior
// MAGI + filing status to its surcharge tier, the per-person and household
// annual cost, and the headroom to the next cliff, via /calc/irmaa.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 });
const money0 = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });

export async function renderIrmaa(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.irmaa.h1.title">// MEDICARE IRMAA SURCHARGE</span></h1>
        <p class="muted small" data-i18n="view.irmaa.hint.intro">
            High-income Medicare beneficiaries pay a surcharge on top of the standard Part B
            and Part D premiums. It's a cliff: cross a MAGI threshold by a single dollar and
            you owe the full higher tier all year — no phase-in. IRMAA is assessed per
            enrolled person on a two-year lookback, so 2026 premiums are set on your 2024
            MAGI. That's why Roth conversions, capital gains, and RMDs taken two years before
            you enroll quietly raise these premiums.
        </p>
        <div class="chart-panel">
            <h2 data-i18n="view.irmaa.h2.inputs">Your income</h2>
            <form id="irmaa-form" class="inline-form">
                <label><span data-i18n="view.irmaa.label.magi">2024 Modified AGI ($)</span>
                    <input type="number" step="0.01" min="0" name="magi_usd" value="150000" required></label>
                <label><span data-i18n="view.irmaa.label.status">Filing status</span>
                    <select name="filing_status">
                        <option value="single" data-i18n="view.irmaa.status.single">Single</option>
                        <option value="married_joint" data-i18n="view.irmaa.status.joint">Married filing jointly</option>
                        <option value="married_separate" data-i18n="view.irmaa.status.separate">Married filing separately</option>
                    </select></label>
                <label data-tip="view.irmaa.tip.both"><input type="checkbox" name="both_spouses_on_medicare"> <span data-i18n="view.irmaa.label.both">Both spouses on Medicare</span></label>
                <button class="primary" type="submit" data-i18n="view.irmaa.btn.run">Find tier</button>
            </form>
        </div>
        <div id="irmaa-result"></div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#irmaa-form');
    form.addEventListener('submit', async (e) => {
        e.preventDefault();
        const fd = new FormData(e.target);
        const body = {
            magi_usd: Number(fd.get('magi_usd')) || 0,
            filing_status: fd.get('filing_status'),
            both_spouses_on_medicare: fd.get('both_spouses_on_medicare') != null,
        };
        try {
            const r = await api.calcIrmaa(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.irmaa.toast.error'), { level: 'error' });
        }
    });
    form.dispatchEvent(new Event('submit'));
}

function renderResult(mount, r) {
    const el = mount.querySelector('#irmaa-result');
    const tierCls = r.tier === 0 ? 'pos' : 'neg';
    const upper = r.upper_threshold_usd == null
        ? t('view.irmaa.no_cap')
        : money0(r.upper_threshold_usd);
    const headroom = r.headroom_to_next_cliff_usd == null
        ? '—'
        : money0(r.headroom_to_next_cliff_usd);
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.irmaa.h2.result">Your surcharge</h2>
            <div class="cards">
                <div class="card ${tierCls}"><div class="label" data-i18n="view.irmaa.card.tier">IRMAA tier</div>
                    <div class="value ${tierCls}">${r.tier} / 5</div></div>
                <div class="card ${r.tier === 0 ? 'pos' : 'neg'}"><div class="label" data-i18n="view.irmaa.card.annual">Annual surcharge (per person)</div>
                    <div class="value">${money0(r.annual_surcharge_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.irmaa.card.household">Household annual surcharge</div>
                    <div class="value">${money0(r.household_annual_surcharge_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.irmaa.card.band">This tier's MAGI band</div>
                    <div class="value">${money0(r.lower_threshold_usd)} – ${upper}</div></div>
                <div class="card"><div class="label" data-i18n="view.irmaa.card.headroom">Headroom to next cliff</div>
                    <div class="value">${headroom}</div></div>
            </div>
            <table class="data-table">
                <thead><tr>
                    <th data-i18n="view.irmaa.col.part">Premium part</th>
                    <th data-i18n="view.irmaa.col.monthly">Monthly</th>
                    <th data-i18n="view.irmaa.col.surcharge">Monthly surcharge</th>
                </tr></thead>
                <tbody>
                    <tr><td data-i18n="view.irmaa.row.part_b">Part B</td>
                        <td>${money(r.part_b_monthly_usd)}</td><td>${money(r.part_b_surcharge_usd)}</td></tr>
                    <tr><td data-i18n="view.irmaa.row.part_d">Part D (added to plan premium)</td>
                        <td>—</td><td>${money(r.part_d_surcharge_usd)}</td></tr>
                    <tr class="emph"><td data-i18n="view.irmaa.row.total">Total surcharge</td>
                        <td>—</td><td class="${r.tier === 0 ? 'pos' : 'neg'}">${money(r.monthly_surcharge_usd)}</td></tr>
                </tbody>
            </table>
            <p class="muted small" data-i18n="view.irmaa.note">2026 premiums (based on 2024 MAGI). Standard Part B is $202.90/mo. IRMAA is charged per enrolled individual; brackets are inflation-adjusted yearly.</p>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/irmaa.rs
//! IRMAA — Medicare Income-Related Monthly Adjustment Amount (2026).
//!
//! High-income Medicare beneficiaries pay a surcharge on top of the
//! standard Part B and Part D premiums. The surcharge is set by a **cliff**
//! schedule: cross a Modified AGI threshold by a single dollar and you owe
//! the full higher tier for the whole year — there is no phase-in. IRMAA is
//! assessed per enrolled individual, on a **two-year lookback** (2026
//! premiums use 2024 MAGI), which is why Roth conversions, capital gains,
//! and RMDs taken two years before age 65/67 quietly raise these premiums.
//!
//! Three filing schedules apply. Married-filing-separately (while living
//! with a spouse) uses a compressed schedule with only the standard tier
//! and the top two surcharge tiers. Figures below are the 2026 amounts
//! (standard Part B premium $202.90/mo); brackets are inflation-adjusted
//! each year. Pure compute.

use serde::{Deserialize, Serialize};

/// The premium year these brackets apply to. IRMAA uses the MAGI from two
/// calendar years earlier (2026 premiums are set on 2024 MAGI).
pub const PREMIUM_YEAR: u16 = 2026;
pub const MAGI_LOOKBACK_YEAR: u16 = 2024;

/// Standard (tier-0) Part B monthly premium for [`PREMIUM_YEAR`].
pub const STANDARD_PART_B_USD: f64 = 202.90;

/// Per-tier Part B total monthly premium (CMS multiples × standard).
const PART_B_MONTHLY: [f64; 6] = [202.90, 284.06, 405.80, 527.54, 649.28, 690.19];
/// Per-tier Part D monthly IRMAA surcharge (added to the plan premium).
const PART_D_SURCHARGE: [f64; 6] = [0.0, 14.50, 37.50, 60.40, 83.30, 91.00];

/// Upper MAGI bound (inclusive) of each tier; the last tier is unbounded.
const SINGLE_UPPER: [f64; 5] = [109_000.0, 137_000.0, 171_000.0, 205_000.0, 500_000.0];
const JOINT_UPPER: [f64; 5] = [218_000.0, 274_000.0, 342_000.0, 410_000.0, 750_000.0];
/// MFS (living with spouse): standard up to $109k, tier 4 up to $391k, then tier 5.
const SEPARATE_STD_UPPER: f64 = 109_000.0;
const SEPARATE_T4_UPPER: f64 = 391_000.0;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FilingStatus {
    Single,
    MarriedJoint,
    MarriedSeparate,
}

#[derive(Debug, Clone, Deserialize)]
pub struct IrmaaInput {
    /// Modified AGI from the lookback year (two years before the premium year).
    pub magi_usd: f64,
    pub filing_status: FilingStatus,
    /// MFJ only: are both spouses enrolled in Medicare? Doubles the household total.
    #[serde(default)]
    pub both_spouses_on_medicare: bool,
}

#[derive(Debug, Clone, Serialize)]
pub struct IrmaaResult {
    /// IRMAA tier 0 (no surcharge) through 5 (highest).
    pub tier: u8,
    /// Lower MAGI bound of the tier (exclusive above tier 0).
    pub lower_threshold_usd: f64,
    /// Upper MAGI bound of the tier; `None` for the unbounded top tier.
    pub upper_threshold_usd: Option<f64>,
    /// MAGI headroom before the next cliff; `None` at the top tier.
    pub headroom_to_next_cliff_usd: Option<f64>,
    /// Total Part B monthly premium at this tier.
    pub part_b_monthly_usd: f64,
    /// Part B surcharge over the standard premium.
    pub part_b_surcharge_usd: f64,
    /// Part D monthly IRMAA surcharge (added to the drug-plan premium).
    pub part_d_surcharge_usd: f64,
    /// Part B + Part D surcharge, per person, per month.
    pub monthly_surcharge_usd: f64,
    /// Per-person annual surcharge.
    pub annual_surcharge_usd: f64,
    /// Household annual surcharge (×2 when MFJ and both are on Medicare).
    pub household_annual_surcharge_usd: f64,
}

/// Resolve (tier, lower bound, upper bound) for a MAGI under a filing status.
fn classify(magi: f64, status: FilingStatus) -> (u8, f64, Option<f64>) {
    match status {
        FilingStatus::Single | FilingStatus::MarriedJoint => {
            let upper = if status == FilingStatus::Single { &SINGLE_UPPER } else { &JOINT_UPPER };
            for (i, &cap) in upper.iter().enumerate() {
                if magi <= cap {
                    let lo = if i == 0 { 0.0 } else { upper[i - 1] };
                    return (i as u8, lo, Some(cap));
                }
            }
            // Above the last finite cap → top tier (index 5), unbounded.
            (5, upper[4], None)
        }
        FilingStatus::MarriedSeparate => {
            if magi <= SEPARATE_STD_UPPER {
                (0, 0.0, Some(SEPARATE_STD_UPPER))
            } else if magi < SEPARATE_T4_UPPER {
                (4, SEPARATE_STD_UPPER, Some(SEPARATE_T4_UPPER))
            } else {
                (5, SEPARATE_T4_UPPER, None)
            }
        }
    }
}

pub fn compute(i: &IrmaaInput) -> IrmaaResult {
    let magi = i.magi_usd.max(0.0);
    let (tier, lower, upper) = classify(magi, i.filing_status);
    let t = tier as usize;

    let part_b_monthly = PART_B_MONTHLY[t];
    let part_b_surcharge = part_b_monthly - STANDARD_PART_B_USD;
    let part_d_surcharge = PART_D_SURCHARGE[t];
    let monthly_surcharge = part_b_surcharge + part_d_surcharge;
    let annual_surcharge = monthly_surcharge * 12.0;

    let multiplier =
        if i.filing_status == FilingStatus::MarriedJoint && i.both_spouses_on_medicare {
            2.0
        } else {
            1.0
        };

    IrmaaResult {
        tier,
        lower_threshold_usd: lower,
        upper_threshold_usd: upper,
        headroom_to_next_cliff_usd: upper.map(|u| (u - magi).max(0.0)),
        part_b_monthly_usd: part_b_monthly,
        part_b_surcharge_usd: part_b_surcharge,
        part_d_surcharge_usd: part_d_surcharge,
        monthly_surcharge_usd: monthly_surcharge,
        annual_surcharge_usd: annual_surcharge,
        household_annual_surcharge_usd: annual_surcharge * multiplier,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn inp(magi: f64, status: FilingStatus) -> IrmaaInput {
        IrmaaInput { magi_usd: magi, filing_status: status, both_spouses_on_medicare: false }
    }

    #[test]
    fn standard_tier_has_no_surcharge() {
        let r = compute(&inp(80_000.0, FilingStatus::Single));
        assert_eq!(r.tier, 0);
        assert!((r.part_b_monthly_usd - 202.90).abs() < 1e-9);
        assert!(r.monthly_surcharge_usd.abs() < 1e-9);
        assert!(r.annual_surcharge_usd.abs() < 1e-9);
    }

    #[test]
    fn single_tier_1_amounts() {
        // $120k single → tier 1: Part B 284.06 (surcharge 81.16) + Part D 14.50.
        let r = compute(&inp(120_000.0, FilingStatus::Single));
        assert_eq!(r.tier, 1);
        assert!((r.part_b_surcharge_usd - 81.16).abs() < 1e-9);
        assert!((r.part_d_surcharge_usd - 14.50).abs() < 1e-9);
        assert!((r.monthly_surcharge_usd - 95.66).abs() < 1e-9);
        assert!((r.annual_surcharge_usd - 95.66 * 12.0).abs() < 1e-9);
    }

    #[test]
    fn cliff_is_inclusive_lower_exclusive_upper() {
        // Exactly at $109k → still tier 0; one dollar over → tier 1.
        assert_eq!(compute(&inp(109_000.0, FilingStatus::Single)).tier, 0);
        assert_eq!(compute(&inp(109_000.01, FilingStatus::Single)).tier, 1);
        // Exactly at $137k → tier 1; just over → tier 2.
        assert_eq!(compute(&inp(137_000.0, FilingStatus::Single)).tier, 1);
        assert_eq!(compute(&inp(137_000.01, FilingStatus::Single)).tier, 2);
    }

    #[test]
    fn joint_thresholds_double_the_single_brackets() {
        // $200k MFJ is under the $218k standard cap → tier 0.
        assert_eq!(compute(&inp(200_000.0, FilingStatus::MarriedJoint)).tier, 0);
        // $250k MFJ → tier 1.
        assert_eq!(compute(&inp(250_000.0, FilingStatus::MarriedJoint)).tier, 1);
    }

    #[test]
    fn joint_household_doubles_when_both_enrolled() {
        let one = compute(&inp(250_000.0, FilingStatus::MarriedJoint));
        let both = compute(&IrmaaInput {
            magi_usd: 250_000.0,
            filing_status: FilingStatus::MarriedJoint,
            both_spouses_on_medicare: true,
        });
        assert!((both.household_annual_surcharge_usd - 2.0 * one.annual_surcharge_usd).abs() < 1e-9);
        // Single-enrolled household equals the per-person amount.
        assert!((one.household_annual_surcharge_usd - one.annual_surcharge_usd).abs() < 1e-9);
    }

    #[test]
    fn top_tier_is_unbounded_with_max_amounts() {
        let r = compute(&inp(600_000.0, FilingStatus::Single));
        assert_eq!(r.tier, 5);
        assert!((r.part_b_monthly_usd - 690.19).abs() < 1e-9);
        assert!((r.part_d_surcharge_usd - 91.00).abs() < 1e-9);
        assert!(r.upper_threshold_usd.is_none());
        assert!(r.headroom_to_next_cliff_usd.is_none());
    }

    #[test]
    fn married_separate_uses_compressed_schedule() {
        // MFS jumps straight from standard to tier 4, then tier 5.
        assert_eq!(compute(&inp(90_000.0, FilingStatus::MarriedSeparate)).tier, 0);
        assert_eq!(compute(&inp(150_000.0, FilingStatus::MarriedSeparate)).tier, 4);
        assert_eq!(compute(&inp(400_000.0, FilingStatus::MarriedSeparate)).tier, 5);
    }

    #[test]
    fn headroom_counts_down_to_the_next_cliff() {
        // $130k single, tier 1 cap $137k → $7k of headroom.
        let r = compute(&inp(130_000.0, FilingStatus::Single));
        assert_eq!(r.tier, 1);
        assert!((r.headroom_to_next_cliff_usd.unwrap() - 7_000.0).abs() < 1e-9);
    }
}
